keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,2248 @@
+@@ -45,5 +45,14 @@ message PodSandboxConfig {
+     LinuxPodSandboxConfig linux = 8;
+     // Optional configurations specific to Windows hosts.
+     WindowsPodSandboxConfig windows = 9;
++    // Configuration of QoS resources.
++    repeated PodQOSResource qos_resources = 10;
+ }
+
++// PodQOSResource specifies the assignment of one QoS resource for a
++// pod.
++message PodQOSResource {
++    // Name of the QoS resource.
++    string name = 1;
++    // Name of the class.
++    string class = 2;
++}
+```
+
+#### PodSandboxStatus
+
+The `PodSandboxStatus` message will be extended to report back assignment of
+pod-level QoS-class resources. The runtime must report back assignment of all
+supported QoS-class resources, also defaults for resources that the client
+didn't request anythin (see [implicit defaults](#implicit-defaults))
+
+```diff
+@@ -545,6 +545,8 @@ message PodSandboxStatus {
+     // runtime configuration used for this PodSandbox.
+     string runtime_handler = 9;
++    // Assignment of QoS resources. The runtime must report back assignment
++    // of all available QoS resources on the node, also possible defaults for
++    // resources that the client didn't explicitly request.
++    repeated PodQOSResource qos_resources = 10;
+ }
+
+```
+
+#### ContainerStatus
+
+The `ContainerResources` message (part of `ContainerStatus`) will be extended
+to report back QoS-class resource assignments of a container, similar to other
+resources. The runtime must report back assignment of all supported QoS-class
+resources, also defaults for resources that the client didn't request
+explicitly (see [implicit defaults](#implicit-defaults))
+
+```diff
+@@ -1251,6 +1269,8 @@ message ContainerResources {
+     LinuxContainerResources linux = 1;
+     // Resource limits configuration specific to Windows container.
+     WindowsContainerResources windows = 2;
++    // Assignment of QoS resources. The runtime must report back assignment
++    // of all available QoS resources on the node, also possible defaults for
++    // resources that the client didn't explicitly request.
++    repeated ContainerQOSResource qos_resources = 3;
+```
+
+#### RuntimeStatus
+
+Extend the `RuntimeStatus` message with new `resources` field that is used to
+communicate the available QoS-class resources from the runtime to the client.
+This information is dynamic i.e. the available QoS-class resources (or their
+properties like available classes or per-class capacity) may change over time,
+e.g. after re-configuration of the runtime.
+
+Names of the QoS-class resources must be unique, also across pod-level and
+container-level resources. I.e. the same name must not be advertised as both a
+pod-level and container-level resource.
+
+```diff
+ message RuntimeStatus {
+     // List of current observed runtime conditions.
+     repeated RuntimeCondition conditions = 1;
++    // Information about the discovered resources
++    ResourcesInfo resources = 2;
```

Comment:
We don't have concept right now in Kubernetes where resources will be "nested" under the RuntimeClass handlers. In overall, it would in my opinion screw up the accounting and resource discovery in multiple components. What we can potentially do later on, as separate KEP, is automatic discovery of the runtime classes/handlers, and "boolean" flags if some runtime handler supports particular resource or not. This way it would be possible to keep accounting global per node, but at the same time report pods as "pending" in scheduler if there is no node with desired combination of runtime handler + set of requested resources (native, extended, qos,....) are available. Simple but user friendly UX IMHO.

Reply:
@mikebrow I put runtime class specific QoS as a non-goal in the KEP. This deserves a separate KEP if we want to have this as I really think that it should then cover other resource management, not just QoS and it will roll like a snowball all over the APIs and the codebase.